File: Data Structures & Algorithms/maximum-subarray/submission-1.java
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums.length == 0){
            return 0;
        }

    int result = Integer.MIN_VALUE;

    for(int i = 0; i < nums.length; i++){ 
        int currSum = 0;
        for(int j = i; j < nums.length; j++){ 
            currSum += nums[j]; 
            result = Math.max(currSum, result);
        }
    }


    return result;





    }
}

File: Data Structures & Algorithms/partition-to-k-equal-sum-subsets/submission-1.py
from functools import cache

class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        total = sum(nums)

        if total % k != 0: 
            return False 


        target = total // k   ## need to divide for an integer 

        @cache
        def dfs(mask, currSum): 
            if mask  == (1 << n) - 1: 
                return True 

            if currSum == target: 
                return dfs(mask, 0)

            for i in range(n): 

                if (mask & (1 << i)) == 0: 
                    if currSum + nums[i] <= target: 
                        attempt = dfs(mask | (1 << i), currSum + nums[i])
                        if attempt: 
                            return True 
                else: 
                    continue 

            return False

        return dfs(0, 0)

        

            

            





